File: app/src/main/java/com/example/retailpulseassignment/mlkit/Classifier.java
package com.example.retailpulseassignment.mlkit;

import android.graphics.Bitmap;

import org.apache.commons.math3.ml.distance.EuclideanDistance;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Map;
import java.util.Objects;

import static com.example.retailpulseassignment.utils.Constants.BATCH_SIZE;
import static com.example.retailpulseassignment.utils.Constants.IMAGE_MEAN;
import static com.example.retailpulseassignment.utils.Constants.IMAGE_STD;
import static com.example.retailpulseassignment.utils.Constants.INPUT_SIZE;
import static com.example.retailpulseassignment.utils.Constants.PIXEL_SIZE;

public class Classifier {

    private static final String TAG = Classifier.class.getSimpleName();
    private int[] mLabels;
    private Map<Integer,double[]> mVectorOutputs;

    public Classifier(int[] mLabels, Map<Integer,double[]> mVectorOutputs) {
        this.mLabels = mLabels;
        this.mVectorOutputs = mVectorOutputs;
    }


    public int getResult(float[] outProbabilities) throws IOException {
        double[] euclideanDistances = new double[32];
        for (int i =0; i < mVectorOutputs.size(); i++){
            euclideanDistances[i] = new EuclideanDistance()
                    .compute(Objects.requireNonNull(mVectorOutputs.get(i)),
                            convertFloatsToDoubles(outProbabilities));
        }
        return getMinEuclideanDistance(euclideanDistances);
    }


    private int getMinEuclideanDistance(double[] euclideanDistances) throws IOException {
        double minDistance = euclideanDistances[0];
        int pos = 0;
        for (int i=0;i<euclideanDistances.length;i++){
            if (minDistance>euclideanDistances[i]){
                minDistance = euclideanDistances[i];
                pos = i;
            }
        }
        return mLabels[pos];
    }


    public static ByteBuffer convertBitmapToByteBuffer(Bitmap bitmap) {
        ByteBuffer byteBuffer = ByteBuffer.allocateDirect(4 * BATCH_SIZE * INPUT_SIZE * INPUT_SIZE * PIXEL_SIZE);
        byteBuffer.order(ByteOrder.nativeOrder());
        int[] intValues = new int[INPUT_SIZE * INPUT_SIZE];
        bitmap.getPixels(intValues, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());
        int pixel = 0;
        for (int i = 0; i < INPUT_SIZE; ++i) {
            for (int j = 0; j < INPUT_SIZE; ++j) {
                final int val = intValues[pixel++];
                byteBuffer.putFloat((((val >> 16) & 0xFF)-IMAGE_MEAN)/IMAGE_STD);
                byteBuffer.putFloat((((val >> 8) & 0xFF)-IMAGE_MEAN)/IMAGE_STD);
                byteBuffer.putFloat((((val) & 0xFF)-IMAGE_MEAN)/IMAGE_STD);
            }
        }
        return byteBuffer;
    }


    public static double[] convertFloatsToDoubles(float[] input) {
        if (input == null) {
            return null; // Or throw an exception - your choice
        }
        double[] output = new double[input.length];
        for (int i = 0; i < input.length; i++) {
            output[i] = input[i];
        }
        return output;
    }

}

File: app/src/main/java/com/rahulshah456/rockpaperscissors/mlkit/Classifier.kt
package com.rahulshah456.rockpaperscissors.mlkit

import android.graphics.Bitmap
import com.rahulshah456.rockpaperscissors.utils.Constants
import org.apache.commons.math3.ml.distance.EuclideanDistance
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

class Classifier(private val mLabels: IntArray, private val mVectorOutputs: Map<Int, DoubleArray>) {


    @Throws(IOException::class)
    fun getResult(outProbabilities: FloatArray?): Int {
        val euclideanDistances = DoubleArray(32)
        for (i in 0 until  mVectorOutputs.size) {
            euclideanDistances[i] = EuclideanDistance()
                    .compute(Objects.requireNonNull(mVectorOutputs[i]),
                            convertFloatsToDoubles(outProbabilities))
        }
        return getMinEuclideanDistance(euclideanDistances)
    }

    @Throws(IOException::class)
    private fun getMinEuclideanDistance(euclideanDistances: DoubleArray): Int {
        var minDistance = euclideanDistances[0]
        var pos = 0
        for (i in euclideanDistances.indices) {
            if (minDistance > euclideanDistances[i]) {
                minDistance = euclideanDistances[i]
                pos = i
            }
        }
        return mLabels[pos]
    }


    companion object {

        @JvmStatic
        fun convertBitmapToByteBuffer(bitmap: Bitmap): ByteBuffer {
            val byteBuffer = ByteBuffer.allocateDirect(4 * Constants.BATCH_SIZE * Constants.INPUT_SIZE * Constants.INPUT_SIZE * Constants.PIXEL_SIZE)
            byteBuffer.order(ByteOrder.nativeOrder())
            val intValues = IntArray(Constants.INPUT_SIZE * Constants.INPUT_SIZE)
            bitmap.getPixels(intValues, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
            var pixel = 0
            for (i in 0 until Constants.INPUT_SIZE) {
                for (j in 0 until Constants.INPUT_SIZE) {
                    val `val` = intValues[pixel++]
                    byteBuffer.putFloat(((`val` shr 16 and 0xFF) - Constants.IMAGE_MEAN) / Constants.IMAGE_STD)
                    byteBuffer.putFloat(((`val` shr 8 and 0xFF) - Constants.IMAGE_MEAN) / Constants.IMAGE_STD)
                    byteBuffer.putFloat(((`val` and 0xFF) - Constants.IMAGE_MEAN) / Constants.IMAGE_STD)
                }
            }
            return byteBuffer
        }

        @JvmStatic
        private fun convertFloatsToDoubles(input: FloatArray?): DoubleArray? {
            if (input == null) {
                return null // Or throw an exception - your choice
            }
            val output = DoubleArray(input.size)
            for (i in input.indices) {
                output[i] = input[i].toDouble()
            }
            return output
        }

    }




}